controllers/networkfence_controller.go

```diff
@@ -113,37 +118,71 @@ func (r *NetworkFenceReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, err
 	}
 
-	// check if the networkfence object is getting deleted, if yes, then send
-	// UnfenceCluster request to the driver.
-	// After the UnfenceCluster request is processed, the finalizer will be removed
-	// so that the networkfence object can be deleted gracefully.
-	if !nwFence.GetDeletionTimestamp().IsZero() {
+	nf := NetworkFenceInstance{
+		reconciler:       r,
+		logger:           logger,
+		instance:         nwFence,
+		controllerClient: client,
+	}
+
+	// check if the networkfence object is getting deleted and handle it.
+	if !nf.instance.GetDeletionTimestamp().IsZero() {
 		if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
 
-			err := r.unfenceClusterNetwork(ctx, nwFence, client, logger)
+			err := nf.removeFinalizerFromNetworkFence(ctx)
```

Comment:
This seems wrong the finalizers should be removed only after unfencing as we need to support the fencing state also? or else in case of an error we might end up deleting the CR but unfencing is not done.

Reply:
Yes, if a fencing CR is deleted after setting its state to "Fenced" and having the set of CIDRs fenced, then the set of CIDRs would be left in a fenced state.

It can be interpreted in 2 ways I think. One is to ensure that the cluster has been unfenced before the resource deletion. Another interpretation can be to say that all it has to do is to ensure its DSoW matches with ASoW and the consumer of the resource is aware of what they are doing with this resource.

From my perspective, I think it would be better to make the reconciler attempt to match DSoW with ASoW instead of assuming that when a resource is deleted, the cluster should be unfenced. 

However, we can introduce an option to say "unfence on delete" which upon being true, will lead to the behavior you suggested. i.e. upon deletion of a NetworkFence CR, the reconciler will first unfence the CIDRs from spec and then deletes the resource. This way, the consumer of this resource would have more control over how they can use the resource.

Let me know your thoughts.

Reply:
>It can be interpreted in 2 ways I think. One is to ensure that the cluster has been unfenced before the resource deletion. Another interpretation can be to say that all it has to do is to ensure its DSoW matches with ASoW and the consumer of the resource is aware of what they are doing with this resource.

am not aware of what is DSoW and ASoW can be please explain.

> From my perspective, I think it would be better to make the reconciler attempt to match DSoW with ASoW instead of assuming that when a resource is deleted, the cluster should be unfenced.

If the resource is deleted why does someone wants to keep the cluster in a fenced state?

>However, we can introduce an option to say "unfence on delete" which upon being true, will lead to the behavior you suggested. i.e. upon deletion of a NetworkFence CR, the reconciler will first unfence the CIDRs from spec and then deletes the resource. This way, the consumer of this resource would have more control over how they can use the resource.

IMO We are complicating it by introducing more options to the users. i would like to do unfence when the CR is deleted and the state is fenced to keep it backward compatible and IMO it's an easy way to do fence/unfence.

Reply:
> > It can be interpreted in 2 ways I think. One is to ensure that the cluster has been unfenced before the resource deletion. Another interpretation can be to say that all it has to do is to ensure its DSoW matches with ASoW and the consumer of the resource is aware of what they are doing with this resource.
> 
> am not aware of what is DSoW and ASoW can be please explain.

Oops. I should have used the complete form.

DSoW: Desired State of World
ASoW: Actual State of World

> 
> > From my perspective, I think it would be better to make the reconciler attempt to match DSoW with ASoW instead of assuming that when a resource is deleted, the cluster should be unfenced.
> 
> If the resource is deleted why does someone wants to keep the cluster in a fenced state?

This is mainly assuming the declarative model of doing things where the reconciler tries to match the Actual State of World (ASOW) with the Desired State of World(DSoW). 

I assume one of the feedbacks given to this PR at [1] was to ensure DSoW meets ASoW before deleting the resource ensure that it has met the DSoW mentioned in the spec. 

I think this example where this behavior of unfencing upon deletion can cause problems.

Lets assume we have 3 kube clusters A, B, C sharing the same storage cluster. Now, A goes down and it is fenced off via B. Now before we can clean up A lets say B has to be decommissioned. And this is where ideally the fencing resource from B has to be migrated to the surviving cluster C along with other things. Here the fencing resource is deleted as part of cleanup of B. Hence even though A should be in a fenced state, it can get unfenced if the deletion from B happens after the migration of the fencing CR to C. And this will be the order of execution (i.e. migrate to C and then clean up B) to ensure smooth migration causing A to be silently get unfenced.

Let me know if I am missing something.

[1] https://github.com/csi-addons/kubernetes-csi-addons/pull/152#discussion_r861995122



> 
> > However, we can introduce an option to say "unfence on delete" which upon being true, will lead to the behavior you suggested. i.e. upon deletion of a NetworkFence CR, the reconciler will first unfence the CIDRs from spec and then deletes the resource. This way, the consumer of this resource would have more control over how they can use the resource.
> 
> IMO We are complicating it by introducing more options to the users. i would like to do unfence when the CR is deleted and the state is fenced to keep it backward compatible and IMO it's an easy way to do fence/unfence.

Reply:
Like I said in a different thread, we should not be taking backward compatibility into consideration here. The CRD has been kept at alpha version so that we can fix the expectations now.

Reply:
>Lets assume we have 3 kube clusters A, B, C sharing the same storage cluster. Now, A goes down and it is fenced off via B. Now before we can clean up A lets say B has to be decommissioned. And this is where ideally the fencing resource from B has to be migrated to the surviving cluster C along with other things. Here the fencing resource is deleted as part of cleanup of B. Hence even though A should be in a fenced state, it can get unfenced if the deletion from B happens after the migration of the fencing CR to C. And this will be the order of execution (i.e. migrate to C and then clean up B) to ensure smooth migration causing A to be silently get unfenced.

For this case, it sounds good :+1: Thanks for explaining it.

I think it's taking some other direction. this is all that I meant, I don't like the idea of asking users to create and delete 2 CRs to fence and unfenced the operation. this is what am looking at

>However, we can introduce an option to say "unfence on delete" which upon being true, will lead to the behavior you suggested. i.e. upon deletion of a NetworkFence CR, the reconciler will first unfence the CIDRs from spec and then deletes the resource. This way, the consumer of this resource would have more control over how they can use the resource.

This is the best option for now to avoid more CR creation and deletion for some users. (we can take this in a followup PR)

Reply:
ACK